sdk/python/test_e2e/test_e2e_train_api.py

```diff
@@ -0,0 +1,205 @@
+                bias="none",
+            ),
+        ),
+        num_workers=num_workers,  # nodes parameter for torchrun command.
+        num_procs_per_worker=1,  # nproc-per-node parameter for torchrun command.
+        resources_per_worker={
+            "gpu": 0,
+            "cpu": 2,
+            "memory": "10G",
+        },
+        storage_config={
+            "size": "10Gi",
+            "access_modes": ["ReadWriteOnce"],
+        },
+    )
+
+    logging.info("---------------------------------------------------------------")
+    logging.info(f"List of created {TRAINING_CLIENT.job_kind}s:")
+    logging.info(TRAINING_CLIENT.list_jobs(job_namespace))
+
+    logging.info("---------------------------------------------------------------")
+    logging.info(f"Training job {JOB_NAME} is running...")
+
+    logging.info("---------------------------------------------------------------")
+    wait_timeout = 60 * 60  # 1 hour.
+    polling_interval = 30  # 30 seconds.
+    start_time = time.time()  # Record the start time
+
+    while True:
+        elapsed_time = time.time() - start_time  # Calculate the elapsed time
+        if elapsed_time > wait_timeout:
+            # Raise a TimeoutError if the job takes too long
+            logging.error(
+                f"Training job {JOB_NAME} exceeded the timeout of {wait_timeout} seconds."
+            )
+            TRAINING_CLIENT.delete_job(JOB_NAME, job_namespace)
+            raise TimeoutError(
+                f"Training job {JOB_NAME} did not complete within the allowed time of "
+                f"{wait_timeout} seconds."
+            )
+
+        # Get the list of pods associated with the job.
+        pod_names = TRAINING_CLIENT.get_job_pod_names(
+            name=JOB_NAME, namespace=job_namespace
+        )
+
+        config.load_kube_config()  # Load kube config to interact with the cluster.
+        v1 = client.CoreV1Api()
+
+        # Iterate over each pod to check its status.
+        for pod_name in pod_names:
+            pod_status = v1.read_namespaced_pod_status(
+                name=pod_name, namespace=job_namespace
+            )
+
+            # Ensure that container_statuses is not None before iterating.
+            if pod_status.status.container_statuses is None:
+                logging.warning(
+                    f"Pod {pod_name} has no container statuses available yet."
+                )
+                continue
+
+            # Check if any container in the pod has been restarted, indicating a previous failure.
+            for container_status in pod_status.status.container_statuses:
+                if container_status.restart_count > 0:
+                    logging.warning(
+                        f"Pod {pod_name} in job {JOB_NAME} has been restarted "
+                        f"{container_status.restart_count} times. Retrieving logs..."
+                    )
+
+                    get_logs_of_master_pod(job_namespace, num_workers)
```

Comment:
instead of logs of master pod, should we get logs of the training job? We can also move it outside of the for loop

Reply:
@deepanker13 I'm not sure if we can directly check the logs of the training job instead of the master pod. Do you know how to check the logs of a job?

Reply:
I found it here: https://github.com/kubeflow/training-operator/blob/e9766d171507ad3772cc04c740acb4d1ee4c544c/sdk/python/test/e2e/utils.py#L87.
Thank you for your suggestions! I will add the e2e test in the "test_e2e_pytorchjob.py", which will check the logs of the training job instead.